### PR TITLE
feat/explains partition and split loading

### PR DESCRIPTION
**Pull request by rudolfix**
### Description
This PR documents and tests two methods of splitting pipeline runs in smaller chunks that may be useful for backfilling:
- partition loading: where source data is partitioned into several ranges that may be independently loaded
- split loading: where source data is split into several sequential ranges using limit item

On top of that a few improvements are added:
- filesystem source follows `sort_order` and may be used for split loading
- `LimitItem` allows to count rows (not only yields/batches)
- several improvements to root_key propagation: (1) simplified implementation (2) allows to use parent_key if nesting level < 2 (3) does not enable root_key on scd2 (4) additional tests and docs

see commit log for more

### CODE REVIEW

File: dlt/common/data_writers/writers.py [hunk 1]
```diff
@@ -823,3 +822,24 @@ def create_import_spec(
 
     spec = DataWriter.class_factory(item_file_format, "object", ALL_WRITERS).writer_spec()
     return spec._replace(data_item_format="file")
+
+
+def count_rows_in_items(item: TDataItems) -> int:
+    """Count total number of rows of `items` which may be
+    * single item
+    * list of single items
+    * list of tables like data frames or
```
Comment (anuunchin) [inline dlt/common/data_writers/writers.py]: ... or table-like object? 👀

File: dlt/sources/filesystem/__init__.py [hunk 2]
```diff
@@ -45,25 +46,47 @@ def readers(  # noqa DOC
         file_glob (str, optional): The filter to apply to the files in glob format. by default lists all files in bucket_url non-recursively
         kwargs: (Optional[Dict[str, Any]], optional): Additional arguments passed to fsspec constructor ie. dict(use_ssl=True) for s3fs
         client_kwargs: (Optional[Dict[str, Any]], optional): Additional arguments passed to underlying fsspec native client ie. dict(verify="public.crt) for botocore
+        incremental (Optional[dlt.sources.incremental[Any]]): defines incremental cursor on listed files, with `modification_date`
```
Comment (anuunchin) [inline dlt/sources/filesystem/__init__.py]:
```suggestion
        incremental (Optional[dlt.sources.incremental[Any]]): Defines incremental cursor on listed files, with `modification_date`
```

File: dlt/sources/filesystem/__init__.py [hunk 3]
```diff
@@ -89,8 +113,10 @@ def filesystem(  # noqa DOC
         files_per_page (int, optional): The number of files to process at once, defaults to 100.
         extract_content (bool, optional): If true, the content of the file will be extracted if
             false it will return a fsspec file, defaults to False.
-        kwargs: (Optional[Dict[str, Any]]): Additional arguments passed to fsspec constructor ie. dict(use_ssl=True) for s3fs
-        client_kwargs: (Optional[Dict[str, Any]]): Additional arguments passed to underlying fsspec native client ie. dict(verify="public.crt) for botocore
+        kwargs (Optional[Dict[str, Any]]): Additional arguments passed to fsspec constructor ie. dict(use_ssl=True) for s3fs
+        client_kwargs (Optional[Dict[str, Any]]): Additional arguments passed to underlying fsspec native client ie. dict(verify="public.crt) for botocore
+        incremental (Optional[dlt.sources.incremental[Any]]): defines incremental cursor on listed files, with `modification_date`
```
Comment (anuunchin) [inline dlt/sources/filesystem/__init__.py]:
```suggestion
        incremental (Optional[dlt.sources.incremental[Any]]): Defines incremental cursor on listed files, with `modification_date`
```

File: dlt/pipeline/drop.py [hunk 4]
```diff
@@ -75,15 +79,15 @@ def _create_modified_state(
...
+def prepare_drop_resources(
     schema: Schema,
     state: TPipelineState,
     resources: Union[Iterable[Union[str, TSimpleRegex]], Union[str, TSimpleRegex]] = (),
     state_paths: jsonpath.TAnyJsonPath = (),
     drop_all: bool = False,
     state_only: bool = False,
-    sources: Optional[Union[Iterable[Union[str, TSimpleRegex]], Union[str, TSimpleRegex]]] = None,
-) -> _DropResult:
+    # sources: Optional[Union[Iterable[Union[str, TSimpleRegex]], Union[str, TSimpleRegex]]] = None,
```
Comment (anuunchin) [inline dlt/pipeline/drop.py:89]: Are we removing the source regex, because we're abandoning the idea of multiple sources per schema? 👀

Comment (anuunchin) [reply dlt/pipeline/drop.py:89]: afair (correctly), the `multiple sources per schema` idea was one of the future todo's in one of the tests

File: dlt/common/libs/pyiceberg.py [hunk 5]
```diff
@@ -179,17 +175,21 @@ def create_table(
...
-        location=table_location,
-    ) as txn:
-        # add partitioning
-        with txn.update_spec() as update_spec:
-            for col in partition_columns:
-                update_spec.add_identity(col)
+    schema = ensure_iceberg_compatible_arrow_schema(schema)
+
+    if partition_columns:
+        # use two staged table create if we have a partition
```
Comment (anuunchin) [inline dlt/common/libs/pyiceberg.py]: I don't understand this comment 👀

Comment (anuunchin) [reply dlt/common/libs/pyiceberg.py]: Does this mean the following?

```
# If the table is partitioned, create it in two steps:
# (1) start a create-table transaction, and (2) add the partition spec before committing
```

File: dlt/common/schema/schema.py [hunk 6]
```diff
@@ -357,22 +357,42 @@ def update_schema(self, schema: "Schema") -> None:
...
+        tables = list(schema.tables.values())
+        for table in tables:
+            if not utils.is_nested_table(table):
+                for chain_table in utils.get_nested_tables(schema._schema_tables, table["name"]):
+                    self.update_table(chain_table)
+
+    def drop_tables(self, table_names: Sequence[str]) -> List[TTableSchema]:
+        """Drops tables from the schema and returns the dropped tables. List of table names
+        must contain all nested tables to tables being dropped.
+        """
```
Comment (anuunchin) [inline dlt/common/schema/schema.py:372]: This will have an effect on my PR with the destructive schema sync - in a good  - thank you 👀

File: dlt/common/data_writers/writers.py (hunk 1 above)
Comment (anuunchin) [reply dlt/common/data_writers/writers.py]: arrow tables it is

File: tests/pipeline/test_arrow_sources.py [hunk 7]
```diff
@@ -707,3 +708,41 @@ def test_replace_or_keep_existing_dlt_load_id(has_dlt_column: bool, add_dlt_load
...
+        ),
+        # edge cases
+        pytest.param(lambda: [], 0, id="empty_list"),
+        pytest.param(
+            lambda: [1, pd.DataFrame({"a": [0, 1]})],
+            2,  # falls back to len(list) because first item has no .shape
+            id="mixed_first_scalar",
+        ),
+    ],
+)
```
Comment (anuunchin) [inline tests/pipeline/test_arrow_sources.py:750]: since we're defining an edge case with
```
[1, pd.DataFrame({"a": [0, 1]})]
```

would it make sense to also have a fallback for the following? 👀 

```
[pd.DataFrame({"a": [0, 1]}), 1]
```

File: docs/website/docs/general-usage/merge-loading.md
Comment (anuunchin) [inline docs/website/docs/general-usage/merge-loading.md:1]: Not specific to this file - it seems we need to run the docs with a grammar checker 👀  and align consistency in terms of wording/punctuation

File: docs/website/docs/dlt-ecosystem/verified-sources/rest_api/advanced.md [hunk 8]
```diff
@@ -151,7 +151,7 @@ This is an experimental feature and may change in future releases.
   - a list of callables, each accepting and returning the response object.
 
 
-#### Example A
+### Example A
```
Comment (anuunchin) [inline docs/website/docs/dlt-ecosystem/verified-sources/rest_api/advanced.md]: Are the examples not within the `### Response action` subsection ? 👀

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database/advanced.md [hunk 9]
```diff
@@ -10,12 +10,106 @@ import Header from '../_source-info-header.md';
...
+    print(pipeline.run(table_partition))
+
+# run pipeline incrementally with the max day as initial_value and open range
+incremental_table = sql_table(
+    table="chat_message",
+    incremental=dlt.sources.incremental(
+        "updated_at",
+        initial_value=date_ranges[-1]["end"],  # use the last day as the starting point
+        range_start="open",  # exclude start boundary as it was loaded by last partition
+    ),
```
Comment (anuunchin) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database/advanced.md]: the comment came off as a bit unintuitive for me during the first read, perhaps something like

```
# skip the last partition’s end value by using an open start boundary as it was loaded by the last partition
```

would be clearer 👀

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database/advanced.md [hunk 10]
```diff
@@ -10,12 +10,106 @@ import Header from '../_source-info-header.md';
...
+    incremental=dlt.sources.incremental(
+        "id",  # in most cases you'll use cursor column of timestamp type
+        row_order="asc",  # critical to set row_order when doing split loading
+        range_start="open",  # use open range to disable deduplication
+    ),
+)
+
+# we'll load 2 pages, 1000 rows max on each run
+while not pipeline.run(incremental_table.add_limit(2)).is_empty:
+    pass
```
Comment (anuunchin) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database/advanced.md:103]: this is so cool!

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database/configuration.md [hunk 11]
```diff
@@ -144,17 +144,18 @@ will create `sql_database` folder with the source code that you can import and u
...
+2. **Set an initial value(optional)**: Choose an initial value for the cursor to begin loading data. This could be a specific timestamp or ID from which you wish to start loading. After first run it will be replaced with the maximum cursor value from the selected rows.
+3. **Set the comparison direction in the query**. By default greater than or equal op (**>=**) is used to compare initial/previous value with row column value. You can change it with `last_value_func` argument (**max**/**min**).
+4. **Set if the comparison is inclusive or exclusive**. By default the range is closed (equal values are included). [Look here for explanation and examples](advanced.md#inclusive-and-exclusive-filtering). Note that for closed ranges `dlt` will use [internal deduplication](../../../general-usage/incremental/cursor.md#deduplicate-overlapping-ranges) which adds some processing cost.
+4. **Configure backfill options(optional)**. You can use `end_value` with `range_end` to read data from specified range. You can also control **order returned rows**
+to split long incremental loading into many chunks by time and row count. [Look here for details and examples]
```
Comment (anuunchin) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database/configuration.md:158]: It seems like the part with `order returned rows` should be a separate point, or no? 👀

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database/troubleshooting.md [hunk 12]
```diff
@@ -25,6 +25,14 @@ modify the local pipeline state (after syncing with destination) to add/remove t
 
 ## Troubleshooting connection
 
+#### Pipeline state grows extremely large or I get deduplication state warnings when using incremental
```
Comment (anuunchin) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database/troubleshooting.md]: I think this should be one  ### instead of #### and something shorter - maybe `Handling large pipeline state or deduplication warnings with incremental` - to align with the other subsection names here that start with ing-form verbs

File: docs/website/docs/general-usage/merge-loading.md [hunk 13]
```diff
@@ -218,9 +218,31 @@ def resource():
...
+2. or, if you are sure that nested tables are max 1 level deep: [Explicitly disable root key propagation](#disable-root-key-propagation)
+
+If you try to switch to merge after nested tables were already created you'll get a warning and NULL column violation from your
+destination. You can fix your pipeline by:
+1. Drop affected resources using `dlt pipeline ... drop` command or by using `refresh` argument on the pipeline. This will drop 
+data from related resources and reset the schema so NOT NULL columns can be created.
+2. If you have nested tables up to 1 nesting level you may [Explicitly disable root key propagation](#disable-root-key-propagation)
+3. You can fix your nested tables in both staging and final datasets. Add `_dlt_root_id` to all nested tables and copy data
+from related [root (top level) tables](../general-usage/schema.md#nested-references-root-and-nested-tables) `_dlt_id` (`row_key`).
+In that case `dlt` will update pipeline schema but will skip database migration.
```
Comment (anuunchin) [inline docs/website/docs/general-usage/merge-loading.md:241]: But as a general rule, should we be endorsing approaches that involve manual `tampering` inside the destination outside of dlt's scope in cases like these? (i just feel like the first two approaches are more dlt- idiomatic 👀 )

**Review by anuunchin — CHANGES_REQUESTED** [review]
🧠

File: dlt/pipeline/drop.py (hunk 4 above)
Comment (rudolfix) [reply dlt/pipeline/drop.py:89]: this code had no effect. right now there's 1:1 between schema and a source. so selecting many resources when schema is defined does not make sense

File: tests/pipeline/test_arrow_sources.py (hunk 7 above)
Comment (rudolfix) [reply tests/pipeline/test_arrow_sources.py:750]: right! and the test will fail here

File: docs/website/docs/general-usage/merge-loading.md (hunk 13 above)
Comment (rudolfix) [reply docs/website/docs/general-usage/merge-loading.md:241]: right! but sometimes you want to keep your data and (2) does not apply. it would be cool to have a tool that propagates root key post factum. but we do not have it :)
